include/envoy/event/dispatcher.h

```diff
@@ -41,20 +41,26 @@ class Dispatcher {
   /**
    * Create a client connection.
    * @param address supplies the address to connect to.
+   * @source_address supplies an optional address to bind to.
    * @return Network::ClientConnectionPtr a client connection that is owned by the caller.
    */
   virtual Network::ClientConnectionPtr
-  createClientConnection(Network::Address::InstanceConstSharedPtr address) PURE;
+  createClientConnection(Network::Address::InstanceConstSharedPtr address,
+                         Optional<Network::Address::InstanceConstSharedPtr> source_address =
```

Comment:
I'd probably use the null/empty shared pointer rather than Optional. Essentially shared_ptr already has the optional nature.